examples/mongocxx/causal_consistency.cpp

```diff
@@ -0,0 +1,121 @@
+using bsoncxx::builder::stream::close_document;
+using bsoncxx::builder::stream::finalize;
+
+int main() {
+    using namespace mongocxx;
+
+    instance inst{};
+    client client{mongocxx::uri{"mongodb://localhost/?replicaSet=replset"}};
+
+    // Start Causal Consistency Example 1
+
+    write_concern wc_majority{};
+    wc_majority.acknowledge_level(write_concern::level::k_majority);
+    wc_majority.majority(std::chrono::milliseconds(100));
+
+    read_concern rc_majority{};
+    rc_majority.acknowledge_level(read_concern::level::k_majority);
+
+    // Use a causally-consistent session to run some operations.
+    options::client_session session_opts;
+    session_opts.causal_consistency(true);
+
+    auto session_1 = client.start_session(std::move(session_opts));
+
+    bsoncxx::types::b_timestamp current_date;
+    auto time_point = std::chrono::system_clock::now();
+    std::time_t now = std::chrono::system_clock::to_time_t(time_point);
+    current_date.timestamp = static_cast<uint32_t>(now);
+
+    auto items = client["test"]["items"];
+    items.write_concern(wc_majority);
+    items.read_concern(rc_majority);
+
+    // Run an update_one with our causally-consistent session.
+    auto none = bsoncxx::types::b_null{};
+    auto update_filter = document{} << "sku"
+                                    << "111"
+                                    << "end" << none << finalize;
+    auto update_op = document{} << "$set" << open_document << "end" << current_date
+                                << close_document << finalize;
+    items.update_one(session_1, std::move(update_filter), std::move(update_op), {});
+
+    // Run an insert with our causally-consistent session.
+    auto insert_doc = document{} << "sku"
+                                 << "nuts-111"
+                                 << "name"
+                                 << "Pecans"
+                                 << "start" << current_date << finalize;
+    items.insert_one(session_1, std::move(insert_doc));
+
+    // End Causal Consistency Example 1
+
+    // Start Causal Consistency Example 2
+
+    // Make a new session, session_2, and make it causally-consistent
+    // with session_1, so that session_2 will read session_1's writes.
+    options::client_session session_opts_2;
+    session_opts_2.causal_consistency(true);
+
+    auto session_2 = client.start_session(std::move(session_opts_2));
+
+    // Set the cluster time for session_2 to session_1's cluster time,
+    // and set the operation time for session2 to session2's operation time.
+    auto cluster_time_1 = session_1.cluster_time();
+    auto operation_time_1 = session_1.operation_time();
+    session_2.advance_cluster_time(std::move(cluster_time_1));
+    session_2.advance_operation_time(std::move(operation_time_1));
+
+    read_preference rp_secondary;
+    rp_secondary.mode(read_preference::read_mode::k_secondary);
+    wc_majority.majority(std::chrono::milliseconds(1000));
```

Comment:
This can reuse the `wc_majority` as-is without modifying it.

Reply:
Done.